RedisGraph protocol v2 --compact

paegun opened the pull request:
* Maintain support for RedisGraph protocol v1.
* Add support for RedisGraph protocol v2.
  * locally cache labels, relationshipTypes, propertyKeys.
  * use --compact form for all graph queries.

Should help w/, if not fully address, #1

paegun commented:
I'll compare w/ the Python client over the course of the coming week, but wanted to first attempt an unbreak using the client guidance doc and slight modifications (addition only) of existing tests.

jeffreylovitz commented on spec/redisgraph_quickstart_spec.rb:
```diff
@@ -0,0 +1,71 @@
+# based on queries extracted from
+describe RedisGraph do
+  before(:all) do
+    @r = RedisGraph.new('quickstart')
```
Don't we need a `begin` to match the `rescue`?

paegun replied:
Effectively wrt exception semantics `def` is a `begin`. The same is not true for blocks using `{}` or `do`. This encourages the use of functions.

jeffreylovitz commented on spec/redisgraph_quickstart_spec.rb:
```diff
@@ -0,0 +1,71 @@
+
+    # not in the quickstart, but demonstrates multiple rows
+    it 'should query relations, without a predicate' do
+      q = "MATCH (r:Rider)-[:rides]->(t:Team) RETURN r.name, t.name"
```
Add an ORDER clause to this query to guarantee deterministic results:
`q = "MATCH (r:Rider)-[:rides]->(t:Team) RETURN r.name, t.name ORDER BY r.name"`

paegun replied:
Will do. Great call.

paegun replied:
Added the ORDER BY clause

jeffreylovitz commented on lib/redisgraph.rb:
```diff
@@ -14,31 +14,88 @@ class RedisGraph
+      # defers first read and is invalidated when changed.
+      @labels_proc =  -> { call_procedure('db.labels') }
+      @property_keys_proc = -> { call_procedure('db.propertyKeys') }
+      @relationship_types_proc = -> { call_procedure('db.relationshipTypes') }
```
This is a nice approach!

paegun replied:
Thank you. This RedisGraph::Metadata (name so far in my head) will eventually be factored into a class b/c otherwise I've created a circular dependency between RedisGraph and QueryResult.

paegun replied:
Refactored into RedisGraph::Metadata and pass that rather than the graph into QueryResult, so broke the otherwise circular dependency.

jeffreylovitz commented on lib/redisgraph.rb:
```diff
@@ -14,31 +14,88 @@ class RedisGraph
+    raise QueryError, e
+  end
+
+  def query_type(command)
```
This logic here is not quite correct, as many valid queries will not begin with a MATCH or CREATE clause:
https://s3.amazonaws.com/artifacts.opencypher.org/railroad/SinglePartQuery.html

Also, queries of forms like `CREATE...RETURN` are valid, so this isn't sufficient introspection to determine how the result will be structured. You could backtrack from the end of the query looking for a `RETURN` token, but I think it's easier to just check the length of the received response array.

paegun replied:
That is a huge help. I mostly wanted to ensure to have cache invalidation in place. Will use the response b/c that reuses the handling of the module.

paegun replied:
Removed parsing query type, using the response shape and enum values w/i arrays as tuples to determine the type.

Also removed this eager load of Metadata and instead load on observing a metadata label, propertyType, or relationshipType w/i the parsing of the result set, as suggested w/i the client developer guidance.

jeffreylovitz commented on lib/redisgraph/query_result.rb:
```diff
@@ -19,13 +50,91 @@ def parse_resultset(response)
 
+  def parse_resultset_v2(response)
+    # In the v2 protocol, CREATE does not contain an empty row preceding statistics
+    case @query_type
```
As mentioned at https://github.com/RedisGraph/redisgraph-rb/pull/4/files#r321789944 :
If `response` is of length 1, it's just statistics; if it's of length 3, it also contains a header and 0 or more records.

paegun replied:
Fixed

jeffreylovitz commented on lib/redisgraph/query_result.rb:
```diff
@@ -19,13 +50,91 @@ def parse_resultset(response)
+    # precisely in order and naming to the RETURN clause of the query.
+    header = response[0].map { |_type, el| el }.to_a
+
+    @columns = header.reduce([]) do |agg, it|
```
`_type` here will correspond to the `ColumnType` enum:
https://github.com/RedisGraph/RedisGraph/blob/master/src/resultset/formatters/resultset_formatter.h#L14-L19

It would be safer to have different decoding routines for these types rather than introspecting on the contents of the individual strings.

paegun replied:
Fixed.

jeffreylovitz commented on lib/redisgraph/query_result.rb:
```diff
@@ -19,13 +50,91 @@ def parse_resultset(response)
+
+    # Second row is the actual data returned by the query
+    data = response[1].map do |row|
+      if row.length == 1
```
I think that this conditional is intended to be differentiating between scalar values and node/edge values, but it's actually checking how many columns there are in the row.

Given the graph constructed by:
`CREATE (:person {name: 'Jim', age: 29})`

And running the query:
`MATCH (a) RETURN a`

At this point, `row` == ``[[0, [0], [[0, 2, "Jim"], [1, 3, 29]]]]``, which actually has length 1.

`MATCH (a) RETURN a, a` works as intended.

paegun replied:
Yes and sorry. I've locally already realized how over-complicated I made this portion, finally realizing that, as you said, scalar or columnar are the value classes here.

paegun replied:
Fixed

paegun commented:
@jeffreylovitz Thank you for the review and course corrections.

wrt "Also, I appreciate the switch between v1 and v2 decoding, but I don't think maintaining backwards compatibility is strictly necessary; you can feel free to drop support for v1 if you prefer!", the primary reason I'm maintaining the compatibility is to more easily ensure the v2 version is correct in the semantic result. Being far less familiar w/ the module's internals and protocol, the v1 results are super informative.
When this PR is in good order, I'll likely drop the backwards compatibility and instead provide a brief section in the README on which version of the gem provides support for the previous module version (w/ a nudge to upgrade).

jeffreylovitz commented:
The progress here is good!

There are a few assumptions being made here that aren't borne out by the way RedisGraph emits results.

Your iterator over the header row (https://github.com/RedisGraph/redisgraph-rb/pull/4/files#diff-424ae2a6fc21f2c6476e144300435fc5R77) breaks down the columns perfectly with respect to the `(ColumnType, column name)` entries. Afterwards, it seems that you're trying to determine whether the cache of property keys is out of date by introspecting on the first row of the results.

This second step does not quite work, as each individual node/edge may possess any properties. The last result can conceivably introduce a property with an ID greater than any that has been seen before. This doesn't invalidate any previous results, but will still require a cache update!

Further, this makes the tabular print approach in use now a bit harder to implement, as you'll have to read all results in order to determine which fields to print. I think it would likely be easier to print properties as a map of key-value pairs within each column that represents either a node or edge (scalars are fine). This is how the Python client operates (using the dataset built with the `python demo/social/social_demo.py --debug`):

```
>>> import redis
>>> from redisgraph import Graph
>>> r = redis.Redis(host='localhost', port=6379)
>>> redis_graph = Graph('social', r)
>>> query = "MATCH (a) RETURN a LIMIT 3"
>>> result = redis_graph.query(query)
>>> result.pretty_print()
+-------------------------------------------------------------------+
|                                 a                                 |
+-------------------------------------------------------------------+
|                           {name:"China"}                          |
| {status:"married",gender:"female",age:31,name:"Jane Chernomorin"} |
|                          {name:"Andora"}                          |
+-------------------------------------------------------------------+

internal execution time 0.529493
```

Running the same on this PR, we encounter printing issues because only one property was found by introspecting on the first row:
```
+------------------+---------+--------+----+
| a.name           |
+------------------+---------+--------+----+
| China            |
| Jane Chernomorin | married | female | 31 |
| Andora           |
+------------------+---------+--------+----+
```
Additionally, I have issues running rspec at present:
```
$ rspec

An error occurred while loading ./spec/redisgraph_quickstart_spec.rb.
Failure/Error: __send__(method, file)

SyntaxError:
  /home/tower-linux/dev/redis/clients/redisgraph-rb/spec/redisgraph_quickstart_spec.rb:10: syntax error, unexpected keyword_rescue, expecting keyword_end
    rescue Redis::BaseError => e
          ^
  /home/tower-linux/dev/redis/clients/redisgraph-rb/spec/redisgraph_quickstart_spec.rb:10: syntax error, unexpected =>
    rescue Redis::BaseError => e
                              ^
  /home/tower-linux/dev/redis/clients/redisgraph-rb/spec/redisgraph_quickstart_spec.rb:15: syntax error, unexpected keyword_do_block, expecting keyword_end
  /home/tower-linux/dev/redis/clients/redisgraph-rb/spec/redisgraph_quickstart_spec.rb:71: syntax error, unexpected keyword_end, expecting end-of-input
# /home/tower-linux/dev/redis/clients/redisgraph-rb/vendor/bundle/ruby/2.3.0/gems/rspec-core-3.8.0/lib/rspec/core/configuration.rb:2034:in `load'
```
This is resolved for me by changing the `before(:all)` blocks in the specs to use a `begin...rescue...end` construction. You had mentioned before that this should be unnecessary, so I may be doing something wrong?

As one final note, I wrote the current documentation for the client spec - if you're not finding it sufficiently clear, please let me know how it could be improved!

paegun commented:
@jeffreylovitz I believe I've addressed your feedback wrt columns and result_set data.

Based on the existing specs, including adding specs for what is w/i the quickstart, the results feel right. The edge.weight property `7.8` is emitted as a float ensuring scalar type mapping is in place. I did not feel Edge/Node classes for the array of property key/values is surely necessary/right.


Considering that the failure on my part is/was due to a non-existent test/spec, I feel that this would be the biggest change I'd ask for w/i the client spec, more specs as in rspecs :)

My misunderstanding was primarily due to the RDBMS and Columnar NoSQL store (and others, though not arguing this is the correct) meaning of result_set as a set of rows which contain columns (so a grid). This is/was a bad assumption.

When reading the client spec attempting to validate whether I was reading the --compact result correctly. I appreciate why you didn't use end-of-line comments to annotate. Retaining the output as is is a good idea, BUT, if the output has line numbers, we can refer to the line numbers for annotations. And, the `print_resultset` / `pretty_print` was super helpful.

jeffreylovitz commented on lib/redisgraph/query_result.rb:
```diff
@@ -11,49 +28,105 @@ def print_resultset
+                 return val == "true"
+               when 5 # double
+                 :to_f
+               end
```
We should also have a `when 1` case for `nil` handling.

NULL values can't be encoded as property values, but they're perfectly valid as scalar returns - most typically as the result of a function call, but also with more contrived queries like:
`UNWIND [1, 1.5, null, 'strval', true, false] AS a RETURN a`
or simply:
`RETURN NULL`

paegun replied:
Thank you. Added using the `UNWIND ...` as a test case.

jeffreylovitz commented:
This looks great! I added  one comment to address, then I think we're merge-ready.

You're correct that the Edge/Node cases don't currently need decoding routines when encountered as scalar values. This will be necessary in the future, when we introduce support for returning paths.

Additionally, we added support for array types just a few days ago (https://github.com/RedisGraph/RedisGraph/commit/d58b291f3078275efc9b84ac9be0763d05483350). If you'd like to add handling for that in this PR, this is how it's handled in the Python client for reference:
https://github.com/RedisGraph/redisgraph-py/pull/43/files#diff-13010f37a0a3fc843dbfd06d2793972dR149-R153
I'm happy to add that following the merge of this PR if you'd prefer, however!

Point taken about the lack of specs, it's a definite weakness of this client (and a great difficulty for contributors) that I will try to allocate some time to soon. And thanks for the comments on documentation! I'll make updates accordingly - annotations based on line numbers is an excellent idea.

paegun commented:
@jeffreylovitz I've added support for NULL and TODO'd the _should_ work Array support.
